### PR TITLE
Refactor Tooltip and Progress

### LINKED ISSUE #310: pin specific quests
**Is your feature request related to a problem? Please describe.**
Keeping track of which characters I've done the weekly quest/brawl/etc is currently annoying. similarly world event quests like the brewfest stuff.

**Describe the solution you'd like**
I'd like to be able to pin specific quests, weekly or otherwise, to show as their own item in the list, presumably which would vanish when no longer available for time-limited ones like brawls.

**Describe alternatives you've considered**
I'm aware you can see weeklies by hovering on the weekly quest row, but that's faffy to compare multiple characters with.

### LINKED ISSUE #390: Adding custom daily / weekly quests tracking
Hi,

Adding a way to track a custom daily or weekly quests would be awesome.

For example, I'm trying to get the achievement [Most Minis Wins](https://www.wowhead.com/achievement=13708/most-minis-wins).
Best way to do that completing 4 hidden weeklies: 57135 Beastbot kill, 57134 Recycling quest, 57132 Reclamation Rig and Mechanized Chest (which is already tracked by SavedInstances).

This is just one example out of many possible ways that a custom weekly or daily tracking would be useful.

I've added manually some of these quests when I need them but they are obviously deleted when updating the addon.

Have been using this addon for years, thank you for all the hard work!

### LINKED ISSUE #415: Add daily rare bosses for mounts at demand ingame.
**Is your feature request related to a problem? Please describe.**
It is not related to a problem.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

Hey, I've been adding soundless & arachnoid harvester, which are daily reset bosses with ID into the file progress to track which alts killed or not killed for the day, it is helpful if you do these two bosses on alot of alts to know which one didnt know without logging them, but thats a hassle to do everytime with an update, I would love to be able to add any daily rare that drop toy/mount or whatever from the ingame options.

This is what I do in the progress file.

-- Rares
  {
    name = "Soundless",
    quest = 56298,
    daily = true,
  },
  {
    name = "Arachnoid_Harvester",
    quest = 55512,
    daily = true,
  },

### LINKED ISSUE #445: More support to The Great Vault
* M+ weekly run record
* Raid reward level

### LINKED ISSUE #471: Weekly Quest Tracking of Shadowlands-activities
I'm Having the problem that the weekly Oribos PvP-Quests are not tracked consistently.

For Example https://wowhead.com/quest=62637 is not tracked. (That's the one from the bonus weekly event, near the m+vault)

https://wowhead.com/quest=62285 This one is tracked. (Weekly non rated PvP from near the PvP Vendor)

https://wowhead.com/quest=62288 This one is not tracked (Weekly rated PvP Quest[RBG])

So only from this week i would assume, tracking of the rated PVP quest, and the bonus event quest doesn't work. [Edit: Im EU, so  weekly reset is tomorrow for me to stop confusion]

I also find it a bit hard to see on a glance whats left to do this week, because there are 7 weekly and only the ones done are shown. So you need to know all Quests and all names of the quests, that are sometimes hard to differentiate.

I would really love to have a summary of all SL-Weeklys, where you can instantly see whats missing on your toon, like withe Thorgast and the callings. Maybe only like a section with Shadowlands weeklys "x/y" quests done and when you hover it all weeklys are shown greyed out, and the one already done are green, and the ones on which you already have progress are yellow. Maybe differentiate pvp/pve/and weekly event quests.

These are the 7 weeklys im talking about:

2 Trading Favor Quests, Dungeon PVE Quests below the innkeeper.
2 PVP Quests at PvP Vendor. I think its always one for rated and one for nonrated PvP
1 Weekly Eventquest at the great vault, different content every week
2 Renown Quests (souls&anima)  (will go away later in the addon)

I commented on [this issue](https://github.com/SavedInstances/SavedInstances/issues/438) but it was closed directly after, and i think my suggestion on the topic wasnt properly adressed nor read. If its not okay to open another issue for this, then sry.

### LINKED ISSUE #541: Custom text for repeatable quest tracking
Hi,

I've added some custom text to the "Progress" file to track specific Rares/Quests etc but I'm wondering if there's a way to add something similar but to return a yes/checkmark if 1 of any of 4 possible quests has been completed that week. My intention is to add Replenish the Reservoir/Return Lost Souls but they have unique IDs for each covenant.

Example of what I have added:

  {
    name = "Sha of Anger",
    quest = 32099,
    weekly = true,
  },

As an aside it would also be cool to have something for Shaping Fate that shows both current % or a checkmark (if completed that week) similar to how the Callings currently work?

Thanks

### LINKED ISSUE #556: tracking 9.0 and beyond char/account upgrades
**Is your feature request related to a problem? Please describe.**
Bought upgrades for account and chars specific is hard to keep track on, once one manage more then 3 chars. upgrades of question:
char specific maw upgrades bought from Ve'nari wowhead.com/npc=162804
account wide torghast upgrades bought from Ve'nari wowhead.com/npc=162804
char specific 9.1 upgrades bought from Duchess Mynx wowhead.com/npc=179321
char specific 9.1 upgrades bought from Archivist Roh-Suir wowhead.com/npc=178257

by wow system these upgrades seem to be hidden quests which can be tracked to be be done or not. ie:
for item 'Research Report: Relic Examination Techniques' wowhead.com/item=187136, the in game console command /run print(C_QuestLog.IsQuestFlaggedCompleted(64367)) returns true/false on char to show if one char have used it yet.
for convience each one toggle able in manner of savedinstances badges

**Describe the solution you'd like**
An entry for each char specific maw & Korthia upgrade with maybe an X or checked box icon to indicate chars progress on that upgrade, as well as the account wides. these upgrades are exaustively used by bliz to feature 9.0, 9.1+ and propably more to come beyond. it seems that these upgrades are a thing for blizz in shadowlands with more to come in future content more then in expacs before.

**Describe alternatives you've considered**
none yet

**Additional context**
upgrades by Venari:
Animaflow Stabilizer wowhead.com/item=180949 char specific
Soul-Stabilizing Talisman wowhead.com/item=18458 char specific
Sigil of the Unseen wowhead.com/item=184605 char specific
Encased Riftwalker Essence wowhead.com/item=184613 char specific
Extradimensional Pockets wowhead.com/item=184615 account wide
Bangle of Seniority wowhead.com/item=184617 account wide
Rank Insignia: Acquisitionist .wowhead.com/item=184618 account wide
Loupe of Unusual Charm wowhead.com/item=184619 account wide
Vessel of Unfortunate Spirits wowhead.com/item=184620 account wide
Ritual Prism of Fortune wowhead.com/item=184621 account wide
Animated Levitating Chain wowhead.com/item=184653 char specific
Broker Traversal Enhancer wowhead.com/item=184901 account wide

upgrades by Duchess Mynx:
Vault Anima Tracker wowhead.com/item=186453 char specific

upgrades by Archivist Roh-Suir
Research Report: All-Seeing Crystal wowhead.com/item=186714 account wide
Research Report: Ancient Shrines wowhead.com/item=186716 char specific
Research Report: Adaptive Alloys wowhead.com/item=186717 char specific
Treatise: Relics Abound in the Shadowlands wowhead.com/item=186721 char specific
Treatise: The Study of Anima and Harnessing Every Drop wowhead.com/item=186722 char specifc
Research Report: Relic Examination Techniques wowhead.com/item=187136 char specifc
Research Report: First Alloys wowhead.com/item=187138 char specific
Treatise: Recognizing Stygia and its Uses wowhead.com/item=187145 char specific
Treatise: Bonds of Stygia in Mortals wowhead.com/item=187706 char specifc

### LINKED ISSUE #567: Which alts have opened their presents?
I use SavedInstances each month to remember which of my many alts still need to do their Dark Moon profession quests. I'd love to have the same feature work with opening the presents at Xmas. Display them in the same line as DMF quests. (They won't conflict for screen space since DMF is the first week of the month and presents are the last week.)

### LINKED ISSUE #717: Tracking of Timewalking-Event
Please track the progress of Timewalking-Events (Example: Quest-ID: 72725 - A Shrouded Path Through Time)

It is currently displayed under weekly quests (without progress).

### LINKED ISSUE #739: 10.1 New things to track
Can we get options to track some of the big new weekly quests and events as separate lines?

The chest from Fyrakk Assaults 
Killing the Fyrakk Assaults spawned world boss (There's a turn in quest for them, track that maybe?)
Reward level gained from the Researchers event

(Edit: My original request included the Niffen rep quest being added, because Curseforge's install was not actually installing the SavedInstances-10.1.1.zip version)

### LINKED ISSUE #740: [Bug] Weekly Conquest Tracking Not Working
**Describe the bug**
The weekly conquest tracker (e.g. get 1350 conquest for the week) is not working correctly. Shows 0/1250(0/3) - seems to be for an older expansion or patch.

**Do you have an error log of what happened?**
No error logs

**To Reproduce**
1. Earn conquest
2. Check SavedInstances

**Screenshots**
[Link](https://i.imgur.com/FDOTAGF.jpeg)

**Did you try having SavedInstances as the only enabled addon and everything else (especially something like ElvUI) disabled?**
No

**Which version of SavedInstances are you using?**
10.1.2

**Was it working in a previous version? If yes, which was the last good one?**
Not sure

**Additional context**
N/A

### LINKED ISSUE #741: Add the ability to dynamically add quest tracking from within the addon
**Is your feature request related to a problem? Please describe.**
There are quests that come out that I would like to keep track of across all my toons, for instance, in the latest patch (10.1), there is "A Worthy Ally" and "Fighting is Its Own Reward"

**Describe the solution you'd like**
Through the UI, I would like to be able to manually enter a quest into the "monitor" list.  It could be as sophisticated as adding a button the quest log, or as simple as providing input fields in the options area to allow one to enter the pertinent information.

**Describe alternatives you've considered**
Currently editing "Progress.lua" to have them show up.  PITA because after every update the file needs to be updated again.

**Additional context**
This could be extended to add "inventory" currencies, such as the crests for example.

### PR DESCRIPTION
**Pull request by LiangYuxuan**
Progress module is very popular since introduced, but it's old code has some shortcomings.

1. use increasing number as key: unable to remove old entry and unable to reorder entry
2. massive use of custom functions: nearly half of entries contain custom functions, and most of which are not reusable
3. lack of ability of user customize
4. the tooltip code is forced to put in Core.lua

The refactor to progress and tooltip is focus on previous shortcomings, with following changes

#### Tooltip

* Tooltip and Indicator Tooltip utility functions move to separate file to allow them used by other files

#### Progress

* introduce entry types
  * `single` for old style single quest tracking
  * `custom` for old style custom tracking, with tooltip function included
  * `any` is introduced to track the first quest that player done or on
  * `list` is introduced to track a quest list
  * dropped the feature to get questID by faction, by using `any`
* build-in field support to non-`custom` to adjust the entry behavior
  * `persists`: keep the progress on reset if not completed yet (like the old `KeepProgress` helper function)
  * `fullObjective`: show more than first objective, for the quest have more than one objective
  * and more that allow to convert old custom entry but in patterns to `any` or `list`
* new designed config ui
  * use expansion to separate entries
  * allow user to adjust order
  * user defined entry
    * allow users to create their own entry
    * only allow creating `single` now
    * `any` or `list` will be supported once find a way to enter multi questIDs
* introduce new entry
  * Great Vault (Raid) (closes #445 ) 
  * Grate Vault (PvP) (rewrites, closes #740 )
  * Timewalking Weekend Event (closes #717 )
  * Researchers Under Fire, Disciple of Fyrakk, Secured Shipment (closes #739 )
* drop Torghast Weekly as it has no lockout

also include Zaralek Cavern Professions Quest (#742 ), thanks to @chris-merritt 

Here I'll be happy to close a bunch of issues that request feature to *Custom Quest Tracking*

closes #310 closes #390 closes #415 closes #471 closes #541 closes #556 closes #567 closes #741

### CODE REVIEW

File: SavedInstances/Core/Config.lua [hunk 1]
```diff
@@ -198,7 +198,7 @@ function Config:BuildOptions()
         name = L["Show/Hide the SavedInstances tooltip"],
         guiHidden = true,
         type = "execute",
-        func = function() SI:ToggleDetached() end,
+        func = function() SI:GetModule('Tooltip'):ToggleDetached() end,
```
Comment (Stanzilla) [inline SavedInstances/Core/Config.lua:201]: instead of calling GetModule every time I would probably just add a new variable at the top

**Comment by cremor** [comment]
As far as I know it is enough to get the best reward from Researchers Under Fire. But the addon now shows it as "1/4" which suggests I should still get the 3 other rewards. Is this by design?
grafik

**Comment by LiangYuxuan** [comment]
It's intended. [Flaming Shalewing Subject 01](https://www.wowhead.com/item=205204/flaming-shalewing-subject-01#contained-in-item) is contained in all four bags and not only from the first bag weekly.

**Comment by F-Lambda** [comment]
Would it be possible to change back to the previous display if you already have the mount?

**Pull request merged**